server/src/main/java/org/elasticsearch/common/metrics/ExponentialBucketHistogram.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.common.metrics;
+
+import java.util.Arrays;
+import java.util.concurrent.atomic.LongAdder;
+
+/**
+ * A histogram with a fixed number of buckets of exponentially increasing width.
+ * <p>
+ * The bucket boundaries are defined by increasing powers of two, e.g.
+ * <code>
+ *     (-&infin;, 1), [1, 2), [2, 4), [4, 8), ..., [2^({@link #bucketCount}-2), &infin;)
+ * </code>
+ * There are {@link #bucketCount} buckets.
+ */
+public class ExponentialBucketHistogram {
+
+    private final int bucketCount;
+    private final long lastBucketLowerBound;
+
+    public static int[] getBucketUpperBounds(int bucketCount) {
+        int[] bounds = new int[bucketCount - 1];
+        for (int i = 0; i < bounds.length; i++) {
+            bounds[i] = 1 << i;
+        }
+        return bounds;
+    }
+
+    private int getBucket(long observedValue) {
+        if (observedValue <= 0) {
+            return 0;
+        } else if (lastBucketLowerBound <= observedValue) {
+            return bucketCount - 1;
+        } else {
+            return Long.SIZE - Long.numberOfLeadingZeros(observedValue);
+        }
+    }
+
+    private final LongAdder[] buckets;
+
+    public ExponentialBucketHistogram(int bucketCount) {
+        if (bucketCount < 2 || bucketCount > Integer.SIZE) {
+            throw new IllegalArgumentException("Bucket count must be in [2, " + Integer.SIZE + "], got " + bucketCount);
+        }
```

Comment:
The observations are `long` but the bucket bounds are integers, I suppose because the scale of the numbers we've been putting in here are easily contained in the integer range. Perhaps we could revisit this if we need histograms to hold larger values, but that's not the case for the queue latency metric.